File: sdk/android/src/main/java/com/tuya/smart/rnsdk/home/TuyaRoomModule.java
package com.tuya.smart.rnsdk.home;

import com.facebook.react.bridge.Promise;
import com.facebook.react.bridge.ReactApplicationContext;
import com.facebook.react.bridge.ReactContextBaseJavaModule;
import com.facebook.react.bridge.ReactMethod;
import com.facebook.react.bridge.ReadableMap;
import com.facebook.react.bridge.WritableMap;
import com.tuya.smart.home.sdk.TuyaHomeSdk;
import com.tuya.smart.home.sdk.bean.DeviceAndGroupInRoomBean;
import com.tuya.smart.rnsdk.utils.JsonUtils;
import com.tuya.smart.rnsdk.utils.ReactParamsCheck;
import com.tuya.smart.rnsdk.utils.TuyaReactUtils;

import java.util.Arrays;

import javax.annotation.Nonnull;

import static com.tuya.smart.rnsdk.utils.Constant.DEVID;
import static com.tuya.smart.rnsdk.utils.Constant.GROUPID;
import static com.tuya.smart.rnsdk.utils.Constant.LIST;
import static com.tuya.smart.rnsdk.utils.Constant.NAME;
import static com.tuya.smart.rnsdk.utils.Constant.ROOMID;
import static com.tuya.smart.rnsdk.utils.Constant.coverDTL;
import static com.tuya.smart.rnsdk.utils.Constant.getIResultCallback;

public class TuyaRoomModule extends ReactContextBaseJavaModule {
    public TuyaRoomModule(@Nonnull ReactApplicationContext reactContext) {
        super(reactContext);
    }

    @Nonnull
    @Override
    public String getName() {
        return "TuyaRoomModule";
    }

    @ReactMethod
    public void updateRoom(ReadableMap params, final Promise promise) {
        if (ReactParamsCheck.checkParams(Arrays.asList(ROOMID, NAME), params)) {
            TuyaHomeSdk.newRoomInstance(coverDTL(params.getDouble(ROOMID))).updateRoom(params.getString(NAME), getIResultCallback(promise));
        }
    }

    @ReactMethod
    public void addDevice(ReadableMap params, final Promise promise) {
        if (ReactParamsCheck.checkParams(Arrays.asList(ROOMID, DEVID), params)) {
            TuyaHomeSdk.newRoomInstance(coverDTL(params.getDouble(ROOMID))).addDevice(params.getString(DEVID), getIResultCallback(promise));
        }
    }

    @ReactMethod
    public void addGroup(ReadableMap params, final Promise promise) {
        if (ReactParamsCheck.checkParams(Arrays.asList(ROOMID, GROUPID), params)) {
            TuyaHomeSdk.newRoomInstance(coverDTL(params.getDouble(ROOMID))).addGroup(
                    coverDTL(params.getDouble(GROUPID)), getIResultCallback(promise));
        }
    }

    @ReactMethod
    public void removeDevice(ReadableMap params, final Promise promise) {
        if (ReactParamsCheck.checkParams(Arrays.asList(ROOMID, DEVID), params)) {
            TuyaHomeSdk.newRoomInstance(coverDTL(params.getDouble(ROOMID))).removeDevice(params.getString(DEVID), getIResultCallback(promise));
        }
    }

    @ReactMethod
    public void removeGroup(ReadableMap params, final Promise promise) {
        if (ReactParamsCheck.checkParams(Arrays.asList(ROOMID, GROUPID), params)) {
            TuyaHomeSdk.newRoomInstance(coverDTL(params.getDouble(ROOMID))).removeGroup(
                    coverDTL(params.getDouble(GROUPID)), getIResultCallback(promise));
        }
    }

    @ReactMethod
    public void moveDevGroupListFromRoom(ReadableMap params, final Promise promise) {
        if (ReactParamsCheck.checkParams(Arrays.asList(ROOMID, LIST), params)) {
            TuyaHomeSdk.newRoomInstance(coverDTL(params.getDouble(ROOMID))).moveDevGroupListFromRoom(
                    JsonUtils.parseArray(JsonUtils.toString(TuyaReactUtils.parseToList(params.getArray(LIST))), DeviceAndGroupInRoomBean.class), getIResultCallback(promise));
        }
    }

    @ReactMethod
    public void sortDevInRoom(ReadableMap params, final Promise promise) {
        if (ReactParamsCheck.checkParams(Arrays.asList(ROOMID, LIST), params)) {
            TuyaHomeSdk.newRoomInstance(coverDTL(params.getDouble(ROOMID))).sortDevInRoom(
                    JsonUtils.parseArray(JsonUtils.toString(TuyaReactUtils.parseToList(params.getArray(LIST))), DeviceAndGroupInRoomBean.class), getIResultCallback(promise));
        }
    }
}

File: Example/android/tuyasmart-sdk-react-native/src/main/java/com/tuya/smart/rnsdk/home/TuyaRoomModule.kt
package com.tuya.smart.rnsdk.home

import com.facebook.react.bridge.*
import com.tuya.smart.home.sdk.TuyaHomeSdk
import com.tuya.smart.home.sdk.api.ITuyaRoom
import com.tuya.smart.rnsdk.utils.Constant.NAME
import com.tuya.smart.rnsdk.utils.Constant.DEVID
import com.tuya.smart.rnsdk.utils.Constant.GROUPID
import com.tuya.smart.rnsdk.utils.Constant.ROOMID
import com.tuya.smart.rnsdk.utils.Constant.getIResultCallback
import com.tuya.smart.rnsdk.utils.ReactParamsCheck


class TuyaRoomModule(reactContext: ReactApplicationContext?) : ReactContextBaseJavaModule(reactContext) {
    override fun getName(): String {
        return "TuyaRoomModule"
    }

    /* 更新房间名称 */
    @ReactMethod
    fun updateRoom(params: ReadableMap, promise: Promise) {
        if (ReactParamsCheck.checkParams(arrayOf(ROOMID, NAME), params)) {
            getRoomInstance(params.getDouble(ROOMID)).updateRoom(params.getString(NAME), getIResultCallback(promise))
        }
    }

    /* 添加设备 */
    @ReactMethod
    fun addDevice(params: ReadableMap, promise: Promise) {
        if (ReactParamsCheck.checkParams(arrayOf(ROOMID, DEVID), params)) {
            getRoomInstance(params.getDouble(ROOMID)).addDevice(params.getString(DEVID), getIResultCallback(promise))
        }
    }

    /* 删除设备  */
    @ReactMethod
    fun removeDevice(params: ReadableMap, promise: Promise) {
        if (ReactParamsCheck.checkParams(arrayOf(ROOMID, DEVID), params)) {
            getRoomInstance(params.getDouble(ROOMID)).removeDevice(params.getString(DEVID), getIResultCallback(promise))
        }
    }

    /* 删除群组  */
    @ReactMethod
    fun removeGroup(params: ReadableMap, promise: Promise) {
        if (ReactParamsCheck.checkParams(arrayOf(ROOMID, GROUPID), params)) {
            getRoomInstance(params.getDouble(ROOMID)).removeGroup(params.getDouble(GROUPID).toLong(), getIResultCallback(promise))
        }
    }

    /* 添加群组  */
    @ReactMethod
    fun addGroup(params: ReadableMap, promise: Promise) {
        if (ReactParamsCheck.checkParams(arrayOf(ROOMID, GROUPID), params)) {
            getRoomInstance(params.getDouble(ROOMID)).addGroup(params.getDouble(GROUPID).toLong(), getIResultCallback(promise))
        }
    }

    fun getRoomInstance(roomId: Double): ITuyaRoom {
        return TuyaHomeSdk.newRoomInstance(roomId.toLong())
    }

}